### PR TITLE
gh-101696: Make sure immutable types have a valid version tag

### PR DESCRIPTION
**Pull request by erlend-aasland**
Reverts #101697

* Issue: gh-101696

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @markshannon @kumaraditya303

**Comment by erlend-aasland** [comment]
I guess we can skip NEWS, as this is strictly internal.

File: Objects/typeobject.c [hunk 1]
```diff
@@ -229,6 +229,13 @@ _PyType_CheckConsistency(PyTypeObject *type)
         CHECK(type->tp_traverse != NULL);
     }
 
+    if (_PyType_HasFeature(type, Py_TPFLAGS_VALID_VERSION_TAG)) {
+        CHECK(type->tp_version_tag != 0);
+    }
+    else {
+        CHECK(type->tp_version_tag == 0);
+    }
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:237]: This check is worth adding, regardless of the rest of this PR, right?

File: Objects/typeobject.c [hunk 2]
```diff
@@ -6971,6 +6976,12 @@ PyType_Ready(PyTypeObject *type)
         type->tp_flags |= Py_TPFLAGS_IMMUTABLETYPE;
     }
 
+    /* All immutable types must have a static valid version tag */
+    if (type->tp_flags & Py_TPFLAGS_IMMUTABLETYPE) {
+        type->tp_version_tag = next_version_tag++;
+        type->tp_flags |= Py_TPFLAGS_VALID_VERSION_TAG;
+    }
```
Comment (ericsnowcurrently) [inline Objects/typeobject.c:6983]: tl;dr I'm on board with this.

As I mentioned in [gh-101696](https://github.com/python/cpython/issues/101696#issuecomment-1424535319), this is good as it covers non-builtin static types (AKA non-`_Py_TPFLAGS_STATIC_BUILTIN`), e.g. in community extension modules. [^1]  (The switch to `Py_TPFLAGS_IMMUTABLETYPE` seems fine.)

FWIW, the already merged fix may fit a little better with the vague idea (in my head) that we should be special-casing static types in the runtime as little as possible (or otherwise in more explicit ways, like the dedicated `_PyStaticType_Dealloc()`).  In that spirit, rather than modifying `PyType_Ready()`, I was going to suggest we could find a way to apply `_PyStaticType_Dealloc()` to all static types, not just the builtin ones, which might be a good idea regardless.  However, that feels like overkill for fixing this issue.  Plus, we're already special-casing static types here in `PyType_Ready()`.  (Perhaps my mental model is misaligned...)

Bottom line: dropping the original fix and adding the fix you have here is worth it since it handles non-builtin static types too.


[^1]: One *could* argue that non-builtin (non-core) static types should be discouraged (basically, they already are) and we shouldn't bother adding extra maintenance burden--however small--to accommodate them.  However, that's not where the community is at nor will be for a long time and there's no reason to be so hard-line about it, especially in a case like this where the additional maintenance burden is essentially zero.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
LGTM

FWIW, it would be nice if there were a test that covered the original problem from gh-101696 (i.e. fail without the fix and pass with it), whether by crashing or some other means.  Would you mind adding something like that in this PR.

File: Objects/typeobject.c (hunk 1 above)
Comment (erlend-aasland) [reply Objects/typeobject.c:237]: Yes

File: Objects/typeobject.c (hunk 2 above)
Comment (erlend-aasland) [reply Objects/typeobject.c:6983]: Note that this PR does _not_ special-case static types; the condition is for _immutable_ types.

**Comment by erlend-aasland** [comment]
> it would be nice if there were a test that covered the original problem from [gh-101696](https://github.com/python/cpython/issues/101696) (i.e. fail without the fix and pass with it), whether by crashing or some other means. Would you mind adding something like that in this PR.

The original problem was uncovered by `test_finalize_structseq` in `test_embed` in gh-101304. Without Kumar's fix (d9de0792482d2ded364b0c7d2867b97a5da41b12), that particular test segfaulted. With Kumar's fix applied, the test passed; with Kumar's fix reverted and this PR applied, the test also passes.

I could of course try to add a more deterministic test. I'm not quite sure how to provoke that ATM, but I can look into it.

**Comment by erlend-aasland** [comment]
Closing, as per discussion on the issue.
